@here/harp-mapview/lib/VisibleTileSet.ts

```diff
@@ -59,30 +60,32 @@ export class TileEntry {
 }
 
 /**
- * Missing Typedoc
+ * Represents a unique TileKey and the area it takes up on screen.
  */
 class TileKeyEntry {
-    constructor(public tileKey: TileKey, public area: number) {}
+    constructor(public tileKey: TileKey, public offset: number, public area: number) {}
 }
 
 /**
- * Missing Typedoc
+ * Stores Tile's or TileProxy's in an LRUCache
  */
 class DataSourceCache {
-    readonly tileCache: LRUCache<number, Tile>;
+    readonly tileCache: LRUCache<number, Tile | TileProxy>;
     readonly disposedTiles: Tile[] = [];
 
     constructor(options: VisibleTileSetOptions) {
-        this.tileCache = new LRUCache<number, Tile>(options.tileCacheSize);
-        this.tileCache.evictionCallback = (_, tile) => {
-            if (tile.tileLoader !== undefined) {
+        this.tileCache = new LRUCache<number, Tile | TileProxy>(options.tileCacheSize);
+        this.tileCache.evictionCallback = (_, tileOrProxy) => {
+            if (tileOrProxy.tile.tileLoader !== undefined) {
                 // Cancel downloads as early as possible.
-                tile.tileLoader.cancel();
+                tileOrProxy.tile.tileLoader.cancel();
```

Comment:
This will allow proxies to cancel loading of `parent` visible tiles when proxy is being evicted.

Reply:
No, because canEvict in this case would return false (proxies parent tile is still Visible) and prevent the eviction happening.

Reply:
> canEvict in this case would return false

So this means that all proxies will live as long as their parent tile is visible and this is wrong too.
Proxies should have separate and completly independent lifecycles.

Reply:
True, but I don't think this is a big issue, to get this specific behaviour, the user would need to jump a multiple of 360 longitudinal degrees to keep the parent tile visible and have a proxy which is then redundant. However, as soon as this tile is removed (Pan / zoom / rotate), both TileProxy's could be removed. Even when zoomed out, yes, there may be many TileProxy's visible, but they would be removed as soon as the user zooms in.

Or is there a specific use case where this could be a problem?

Reply:
> but I don't think this is a big issue

That worries me so much, because we will rely on very tricky behaviors.
If we would have complete different lifecycle of tile/proxytile this wouldn't be tricky at all.

In general, i don't even know why proxy should reference original `Tile` object at all.
Why shouldn't it just call `dataSource.getTile(t)` on it's own ?

On high levels, cost of decoding and extra geometry is small.

Optimization - i.e reuse of geometry between normal tiles and offset-ed tiles can be added as next step.

Reply:
> On high levels, cost of decoding and extra

On low level, you usually wouldn't even have "original" non-offseted "parent" in cache at all!

Reply:
> That worries me so much, because we will rely on very tricky behaviors.

The idea isn't that particularly complicated, the tile cache has two types, Tiles and TileProxys which reference the Tile. A tile can be evicted, in which case the TileProxy should be evicted at the next convenience, because it is pointing at something useless, it is similar to a std::weak_ptr, we can see if what it is pointing to is still alive or not.

> If we would have complete different lifecycle of tile/proxytile this wouldn't be tricky at all.
> 
> In general, i don't even know why proxy should reference original `Tile` object at all.
> Why shouldn't it just call `dataSource.getTile(t)` on it's own ?

Yes, I had tried this originally, see the comment here: https://github.com/heremaps/harp.gl/pull/264#discussion_r262151610 however it was deemed a "no-go". I also would prefer this simpler approach to avoid this large change, but that might require some further discussion.

> On high levels, cost of decoding and extra geometry is small.

Agreed, but deemed not acceptable as in previous comment.

> Optimization - i.e reuse of geometry between normal tiles and offset-ed tiles can be added as next step.

Ok, given that you call dataSource.getTile(t) twice, how should one tile reuse the geometry of another Tile. This sounds potentially more complicated than the current idea of using a proxy.

Reply:
> > On high levels, cost of decoding and extra
> 
> On low level, you usually wouldn't even have "original" non-offseted "parent" in cache at all!

Not following, if the Tile or TileProxy is visible, then it will be in the cache.